File: app/src/main/java/com/ayan/ritual/render/Fonts.kt
package com.ayan.ritual.render

import android.content.Context
import android.graphics.Typeface
import androidx.core.content.res.ResourcesCompat
import com.ayan.ritual.R

/**
 * Archivo, in the three weights the design uses.
 *
 * One variable TTF ships in the APK; the font-family XMLs pin the `wght` axis
 * so the Canvas renderer and Compose draw the very same instances.
 */
object Fonts {

    private var extraBold: Typeface? = null
    private var semiBold: Typeface? = null
    private var medium: Typeface? = null

    fun load(context: Context) {
        if (extraBold != null) return
        val app = context.applicationContext
        // A missing or unloadable font must never take the widget down with it.
        extraBold = runCatching { ResourcesCompat.getFont(app, R.font.archivo_extrabold) }
            .getOrNull() ?: Typeface.create("sans-serif-black", Typeface.NORMAL)
        semiBold = runCatching { ResourcesCompat.getFont(app, R.font.archivo_semibold) }
            .getOrNull() ?: Typeface.create("sans-serif-medium", Typeface.NORMAL)
        medium = runCatching { ResourcesCompat.getFont(app, R.font.archivo_medium) }
            .getOrNull() ?: Typeface.create("sans-serif-medium", Typeface.NORMAL)
    }

    fun extraBold(): Typeface = extraBold ?: Typeface.create("sans-serif-black", Typeface.NORMAL)
    fun semiBold(): Typeface = semiBold ?: Typeface.create("sans-serif-medium", Typeface.NORMAL)
    fun medium(): Typeface = medium ?: Typeface.create("sans-serif-medium", Typeface.NORMAL)
}

File: ios/Shared/Fonts.swift
import CoreText
import UIKit

/**
 Archivo, in the three weights the design uses.

 One variable TTF ships in the bundle and the `wght` axis is pinned per weight,
 so the Core Graphics renderer and SwiftUI draw the very same instances — the
 iOS equivalent of the `archivo_*.xml` font families on Android. A missing or
 unloadable font must never take the widget down with it, so every path falls
 back to the system face.
 */
enum Fonts {

    private static let weightAxis: Int = 0x77676874  // 'wght'
    private static var registered = false

    /// Registers the bundled TTF with Core Text. Safe to call repeatedly.
    static func register() {
        if registered { return }
        registered = true
        let bundle = Bundle(for: FontsAnchor.self)
        guard let url = bundle.url(forResource: "Archivo", withExtension: "ttf") else { return }
        CTFontManagerRegisterFontsForURL(url as CFURL, .process, nil)
    }

    static func extraBold(_ size: CGFloat) -> UIFont { archivo(size, weight: 800, fallback: .heavy) }
    static func semiBold(_ size: CGFloat) -> UIFont { archivo(size, weight: 600, fallback: .semibold) }
    static func medium(_ size: CGFloat) -> UIFont { archivo(size, weight: 500, fallback: .medium) }

    private static func archivo(_ size: CGFloat, weight: CGFloat, fallback: UIFont.Weight) -> UIFont {
        register()
        let base = UIFont(name: "Archivo", size: size)
            ?? UIFont(name: "Archivo-Regular", size: size)
            ?? UIFont(name: "Archivo-Variable", size: size)
        guard let base else { return .systemFont(ofSize: size, weight: fallback) }
        let descriptor = base.fontDescriptor.addingAttributes([
            UIFontDescriptor.AttributeName(rawValue: kCTFontVariationAttribute as String): [weightAxis: weight]
        ])
        return UIFont(descriptor: descriptor, size: size)
    }
}

/// Only here so `Bundle(for:)` finds whichever bundle this file was built into
/// — the app and the widget extension each carry their own copy of the font.
private final class FontsAnchor {}
